pkg/kubelet/pod_workers.go

```diff
@@ -221,61 +475,310 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+				status, err = p.podCache.GetNewerThan(update.Pod.UID, lastSyncTime)
+			}
+			if err != nil {
+				// This is the legacy event thrown by manage pod loop all other events are now dispatched
+				// from syncPodFn
+				p.recorder.Eventf(update.Pod, v1.EventTypeWarning, events.FailedSync, "error determining status: %v", err)
+				return err
+			}
+
+			// Take the appropriate action (illegal phases are prevented by UpdatePod)
+			switch {
+			case update.UpdateType == kubetypes.SyncPodTerminated:
+				err = p.syncTerminatedPodFn(update.Pod, status)
+
+			case update.UpdateType == kubetypes.SyncPodKill:
+				p.acknowledgeTerminating(update.Pod)
+				var gracePeriod *int64
+				var podStatusFn func(*v1.PodStatus)
+				if update.KillPodOptions != nil {
+					gracePeriod = update.KillPodOptions.PodTerminationGracePeriodSecondsOverride
+					podStatusFn = update.KillPodOptions.PodStatusFunc
+				}
+				err = p.syncTerminatingPodFn(update.Pod, status, update.RunningPod, gracePeriod, podStatusFn)
+
+			default:
+				err = p.syncPodFn(syncPodOptions{
+					mirrorPod:  update.MirrorPod,
+					pod:        update.Pod,
+					podStatus:  status,
+					updateType: update.UpdateType,
+				})
+			}
+
+			lastSyncTime = time.Now()
+			return err
+		}()
+		// notify the call-back function if the operation succeeded or not
+		if update.OnCompleteFunc != nil {
+			update.OnCompleteFunc(err)
+		}
+		if err != nil {
+			// IMPORTANT: we do not log errors here, the syncPodFn is responsible for logging errors
+			klog.ErrorS(err, "Error syncing pod, skipping", "pod", klog.KObj(update.Pod), "podUID", update.Pod.UID)
+		}
+
+		// if we've successfully terminated the pod, we can move to the next phase
+		if err == nil {
+			switch {
+			case update.UpdateType == kubetypes.SyncPodTerminated:
+				p.completeTerminated(update.Pod)
+				klog.V(4).InfoS("DEBUG: Processing pod event done", "pod", klog.KObj(update.Pod), "podUID", update.Pod.UID, "updateType", update.UpdateType)
+				return
+			case update.UpdateType == kubetypes.SyncPodKill:
+				if update.RunningPod != nil {
+					p.completeTerminatingRuntimePod(update.Pod)
+					klog.V(4).InfoS("DEBUG: Processing pod event done", "pod", klog.KObj(update.Pod), "podUID", update.Pod.UID, "updateType", update.UpdateType)
+					return
+				}
+				p.completeTerminating(update.Pod)
+			}
+		}
+		p.completeWork(update.Pod, err)
+		klog.V(4).InfoS("DEBUG: Processing pod event done", "pod", klog.KObj(update.Pod), "podUID", update.Pod.UID, "updateType", update.UpdateType)
+	}
+}
+
+// acknowledgeTerminating sets the terminating flag on the pod status once the pod worker sees
+// the termination state so that other components know no new containers will be started in this
+// pod.
+func (p *podWorkers) acknowledgeTerminating(pod *v1.Pod) {
+	p.podLock.Lock()
+	defer p.podLock.Unlock()
+
+	if status, ok := p.podSyncStatus[pod.UID]; ok && !status.terminatingAt.IsZero() && !status.terminating {
```

Comment:
In what case would `status.terminatingAt.IsZero()` not be zero here?

Reply:
More as a guard against programmer error, but never (could change it to an assertion)